// relay-compiler/lib/printFlowTypes.js
/**
 * Copyright (c) 2013-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * 
 * @providesModule printFlowTypes
 * @format
 */

'use strict';

var generate = require('babel-generator')['default'];

var traverse = require('babel-traverse')['default'];

var _require = require('./RelayFlowParser'),
    RELAY_CLASSIC_MUTATION = _require.RELAY_CLASSIC_MUTATION;

var _require2 = require('./GraphQLSchemaUtils'),
    getRawType = _require2.getRawType;

var _require3 = require('graphql'),
    GraphQLEnumType = _require3.GraphQLEnumType,
    GraphQLList = _require3.GraphQLList,
    GraphQLNonNull = _require3.GraphQLNonNull,
    GraphQLScalarType = _require3.GraphQLScalarType;

var FIELD_BLACKLIST = ['clientMutationId', 'client_mutation_id'];

/**
 * Prints a given Root or Fragment into a Flow type declaration.
 */
function printFlowTypes(node) {
  if (node.kind === 'Root') {
    // for now, only fragments and mutations have flow types
    if (node.operation === 'mutation') {
      var selection = node.selections[0];
      if (selection.kind === 'LinkedField') {
        var argument = selection.args[0];
        var inputIR = require('./transformInputObjectToIR')(argument);

        var response = [];
        if (node.name !== RELAY_CLASSIC_MUTATION) {
          selection.name = node.name + 'Response';
          response = normalize(transform(selection));
        }
        return normalize(transform(inputIR)).concat(response).map(function (n) {
          return generate(n, { flowCommaSeparator: true }).code;
        }).join('\n\n');
      }
    }
  } else if (node.kind === 'Fragment') {
    return normalize(transform(node)).map(function (n) {
      return generate(n, { flowCommaSeparator: true }).code;
    }).join('\n\n');
  }
}

/**
 * Transforms a "root" type (Fragment or LinkedField) into a Flow export type
 * statement.
 */
function transform(node) {
  return require('babel-types').exportNamedDeclaration(require('babel-types').typeAlias(require('babel-types').identifier(node.name), null, require('babel-types').objectTypeAnnotation(node.selections.map(function (selection) {
    return transformSelection(selection);
  }).reduce(function (prev, curr) {
    if (!curr) {
      return prev;
    }
    return prev.concat(curr);
  }, []), null, null)), [], null);
}

function normalize(ast) {
  var normalizedRoots = [ast];

  var findObjectTypeProperty = {
    enter: function enter(path) {
      if (require('babel-types').isObjectTypeProperty(path)) {
        // Ignore object types that are direct children of the root Flow type
        if (require('babel-types').isTypeAlias(path.parent)) {
          return;
        }
        var name = path.node.key.name;

        path.traverse(findObjectTypeAnnotation, {
          name: this.prevName + '_' + name
        });
      }
    },

    noScope: true
  };
  var findObjectTypeAnnotation = {
    enter: function enter(path) {
      if (require('babel-types').isObjectTypeAnnotation(path)) {
        // This has side effects on the path
        path.traverse(findObjectTypeProperty, { prevName: this.name });

        // Add the node of that transformed path to the array
        normalizedRoots.push(require('babel-types').exportNamedDeclaration(require('babel-types').typeAlias(require('babel-types').identifier(this.name), null, path.node), [], null));

        // Replace that path with a reference to the extracted type
        path.replaceWith(require('babel-types').genericTypeAnnotation(require('babel-types').identifier(this.name), null));
      }
    },

    noScope: true
  };

  traverse(ast, findObjectTypeProperty, {}, { prevName: ast.declaration.id.name });

  return normalizedRoots;
}

/**
 * Transforms a ScalarField or LinkedField to a Flow objectTypeProperty.
 * If `forceNull` is true, the selection will *always* be a nullable field.
 */
function transformSelection(node) {
  var forceNull = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;

  if (node.name && FIELD_BLACKLIST.indexOf(node.name) !== -1) {
    return;
  }

  var name = node.alias || node.name && node.name;
  var annotation = void 0;
  var body = void 0;
  switch (node.kind) {
    case 'ScalarField':
      body = wrapNullOrArray(node.type, transformScalarField(getRawType(node.type)));
      if (body.type !== 'NullableTypeAnnotation' && forceNull) {
        body = require('babel-types').nullableTypeAnnotation(body);
      }
      annotation = require('babel-types').objectTypeProperty(require('babel-types').identifier(name), body);
      break;
    case 'LinkedField':
      body = wrapNullOrArray(node.type, transformLinkedField(node));
      if (body.type !== 'NullableTypeAnnotation' && forceNull) {
        body = require('babel-types').nullableTypeAnnotation(body);
      }
      annotation = require('babel-types').objectTypeProperty(require('babel-types').identifier(name), body);
      break;
    case 'Condition':
      annotation = node.selections.map(function (s) {
        return transformSelection(s, true);
      }).reduce(function (prev, curr) {
        if (!curr) {
          return prev;
        }
        return prev.concat(curr);
      }, []);
      break;
    case 'FragmentSpread':
      return; // noop
    case 'InlineFragment':
      return node.selections.map(function (s) {
        return transformSelection(s);
      }).reduce(function (prev, curr) {
        if (!curr) {
          return prev;
        }
        return prev.concat(curr);
      }, []).map(function (s) {
        s.optional = true;
        return s;
      });
    default:
      throw new Error('Unknown Selection type: ' + node.kind);
  }

  if (Array.isArray(annotation)) {
    return annotation.map(function (a) {
      if (!(node.type instanceof GraphQLNonNull)) {
        a.optional = true;
      }
      return a;
    });
  }

  if (!(node.type instanceof GraphQLNonNull)) {
    annotation.optional = true;
  }
  return [annotation];
}

/**
 * Wraps the given type annotation as a Flow Array.
 */
function getArrayTypeAnnotation(typeAnnotation) {
  return require('babel-types').genericTypeAnnotation(require('babel-types').identifier('Array'), require('babel-types').typeParameterInstantiation([typeAnnotation]));
}

/**
 * Recursively wraps the given type with Array or NonNullable nodes until it
 * reaches the root type.
 */
function wrapNullOrArray(type, child) {
  var nullable = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : true;

  if (!type) {
    return child;
  }

  var annotation = void 0;
  if (type instanceof GraphQLNonNull) {
    return wrapNullOrArray(type.ofType, child, false);
  } else if (type instanceof GraphQLList) {
    annotation = getArrayTypeAnnotation(wrapNullOrArray(type.ofType, child));
  } else {
    annotation = child;
  }

  return nullable ? require('babel-types').nullableTypeAnnotation(annotation) : annotation;
}

/**
 * Transforms a LinkedField into a Flow objectTypeAnnotation.
 */
function transformLinkedField(node) {
  var selections = node.selections.map(function (selection) {
    return transformSelection(selection);
  }).reduce(function (prev, curr) {
    if (!curr) {
      return prev;
    }
    return prev.concat(curr);
  }, []);

  // If there are no selections, then we know the only child was a FragmentSpread
  // so we should make this field an 'any' type. (The alternative is an empty
  // object!)
  if (selections.length) {
    return require('babel-types').objectTypeAnnotation(selections, null, null);
  } else {
    return require('babel-types').anyTypeAnnotation();
  }
}

/**
 * Transforms a ScalarField type into its corresponding Flow AST node.
 */
function transformScalarField(type) {
  if (type instanceof GraphQLScalarType) {
    switch (type.name) {
      case 'Color':
      case 'File':
      case 'ID':
      case 'String':
      case 'Url':
        return require('babel-types').stringTypeAnnotation();
      case 'Float':
      case 'Int':
      case 'Time':
        return require('babel-types').numberTypeAnnotation();
      case 'Boolean':
        return require('babel-types').booleanTypeAnnotation();
      default:
        // Fallback to `any` for custom scalar types.
        return require('babel-types').anyTypeAnnotation();
    }
  } else if (type instanceof GraphQLEnumType) {
    var stringLiterals = type.getValues().map(function (_ref) {
      var value = _ref.value;

      var literal = require('babel-types').stringLiteralTypeAnnotation();
      literal.value = value;
      return literal;
    });

    return require('babel-types').unionTypeAnnotation(stringLiterals);
  } else {
    throw new Error('Could not convert from GraphQL type ' + type.toString());
  }
}

module.exports = printFlowTypes;
