// relay-compiler/lib/RelayGenerateRequisiteFieldsTransform.js
/**
 * Copyright (c) 2013-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * 
 * @providesModule RelayGenerateRequisiteFieldsTransform
 * @format
 */

'use strict';

var _extends3 = _interopRequireDefault(require('babel-runtime/helpers/extends'));

var _toConsumableArray3 = _interopRequireDefault(require('babel-runtime/helpers/toConsumableArray'));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _require = require('graphql'),
    assertAbstractType = _require.assertAbstractType,
    assertCompositeType = _require.assertCompositeType,
    assertLeafType = _require.assertLeafType;

var canHaveSelections = require('./GraphQLSchemaUtils').canHaveSelections,
    getRawType = require('./GraphQLSchemaUtils').getRawType,
    hasID = require('./GraphQLSchemaUtils').hasID,
    implementsInterface = require('./GraphQLSchemaUtils').implementsInterface,
    isAbstractType = require('./GraphQLSchemaUtils').isAbstractType,
    mayImplement = require('./GraphQLSchemaUtils').mayImplement;

var TYPENAME_KEY = '__typename';
var ID = 'id';
var ID_TYPE = 'ID';
var NODE_TYPE = 'Node';
var STRING_TYPE = 'String';

/**
 * A transform that adds "requisite" fields to all nodes:
 * - Adds an `id` selection on any `LinkedField` of type that implements `Node`
 *   or has an id field but where there is no unaliased `id` selection.
 * - Adds `__typename` on any `LinkedField` of a union/interface type where
 *   there is no unaliased `__typename` selection.
 */
function transform(context) {
  var documents = context.documents();
  return documents.reduce(function (ctx, node) {
    var transformedNode = transformNode(context, node);
    return ctx.add(transformedNode);
  }, new (require('./RelayCompilerContext'))(context.schema));
}

function transformNode(context, node) {
  var selections = node.selections.map(function (selection) {
    if (selection.kind === 'LinkedField') {
      return transformField(context, selection);
    } else if (selection.kind === 'InlineFragment' || selection.kind === 'Condition') {
      return transformNode(context, selection);
    } else {
      return selection;
    }
  });
  return (0, _extends3['default'])({}, node, {
    selections: selections
  });
}

function transformField(context, field) {
  var transformedNode = transformNode(context, field);
  var type = field.type;

  var generatedSelections = [].concat((0, _toConsumableArray3['default'])(transformedNode.selections));
  var idSelections = generateIDSelections(context, field, field.type);
  if (idSelections) {
    generatedSelections.push.apply(generatedSelections, (0, _toConsumableArray3['default'])(idSelections));
  }
  if (isAbstractType(type) && !hasUnaliasedSelection(field, TYPENAME_KEY)) {
    var stringType = assertLeafType(context.schema.getType(STRING_TYPE));
    generatedSelections.push({
      kind: 'ScalarField',
      alias: null,
      args: [],
      directives: [],
      handles: null,
      metadata: null,
      name: TYPENAME_KEY,
      type: stringType
    });
  }
  var selections = sortSelections(generatedSelections);
  return (0, _extends3['default'])({}, transformedNode, {
    selections: selections
  });
}

/**
 * @internal
 *
 * Returns an array of zero or more selections to fetch `id` depending on the
 * type of the given field:
 * - If the field already has an unaliased `id` field, do nothing
 * - If the field type has an `id` subfield, return an `id` selection
 * - If the field type is abstract, then generate a `... on Node { id }`
 *   fragment if *any* concrete type implements Node. Then generate a
 *   `... on PossibleType { id }` for every concrete type that does *not*
 *   implement `Node`
 */
function generateIDSelections(context, field, type) {
  if (hasUnaliasedSelection(field, ID)) {
    return null;
  }
  var unmodifiedType = assertCompositeType(getRawType(type));
  var generatedSelections = [];
  // Object or  Interface type that has `id` field
  if (canHaveSelections(unmodifiedType) && hasID(context.schema, unmodifiedType)) {
    var idType = assertLeafType(context.schema.getType(ID_TYPE));
    generatedSelections.push({
      kind: 'ScalarField',
      alias: null,
      args: [],
      directives: [],
      handles: null,
      metadata: null,
      name: ID,
      type: idType
    });
  } else if (isAbstractType(unmodifiedType)) {
    // Union or interface: concrete types may implement `Node` or have an `id`
    // field
    var _idType = assertLeafType(context.schema.getType(ID_TYPE));
    if (mayImplement(context.schema, unmodifiedType, NODE_TYPE)) {
      var nodeType = assertCompositeType(context.schema.getType(NODE_TYPE));
      generatedSelections.push(buildIdFragment(nodeType, _idType));
    }
    var abstractType = assertAbstractType(unmodifiedType);
    context.schema.getPossibleTypes(abstractType).forEach(function (possibleType) {
      if (!implementsInterface(possibleType, NODE_TYPE) && hasID(context.schema, possibleType)) {
        generatedSelections.push(buildIdFragment(possibleType, _idType));
      }
    });
  }
  return generatedSelections;
}

/**
 * @internal
 */
function buildIdFragment(fragmentType, idType) {
  return {
    kind: 'InlineFragment',
    directives: [],
    metadata: null,
    typeCondition: fragmentType,
    selections: [{
      kind: 'ScalarField',
      alias: null,
      args: [],
      directives: [],
      handles: null,
      metadata: null,
      name: ID,
      type: idType
    }]
  };
}

/**
 * @internal
 */
function hasUnaliasedSelection(field, fieldName) {
  return field.selections.some(function (selection) {
    return selection.kind === 'ScalarField' && selection.alias == null && selection.name === fieldName;
  });
}

/**
 * @internal
 *
 * For interoperability with classic systems, sort `__typename` first.
 */
function sortSelections(selections) {
  return [].concat((0, _toConsumableArray3['default'])(selections)).sort(function (a, b) {
    return a.kind === 'ScalarField' && a.name === TYPENAME_KEY ? -1 : b.kind === 'ScalarField' && b.name === TYPENAME_KEY ? 1 : 0;
  });
}

module.exports = { transform: transform };
