// relay-compiler/lib/transformInputObjectToIR.js
/**
 * Copyright (c) 2013-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * TODO @joesavona: enable flow
 * @providesModule transformInputObjectToIR
 * @format
 */

'use strict';

var _require = require('./GraphQLSchemaUtils'),
    getRawType = _require.getRawType;

var _require2 = require('graphql'),
    GraphQLEnumType = _require2.GraphQLEnumType,
    GraphQLInputObjectType = _require2.GraphQLInputObjectType,
    GraphQLNonNull = _require2.GraphQLNonNull,
    GraphQLScalarType = _require2.GraphQLScalarType;

/**
 * Transforms a GraphQLInputObjectType to a RelayIR LinkedField.
 */
function transformInputObjectToIR(node) {
  var type = getRawType(node.type);
  var fields = type.getFields();
  // If the node is the root (an Argument), use the name of the type so it is
  // named 'FooBarData' instead of 'input'

  var _ref = node.kind === 'Argument' ? type : node,
      name = _ref.name;

  return {
    alias: null,
    args: [],
    directives: [],
    handles: null,
    kind: 'LinkedField',
    metadata: null,
    name: name,
    selections: Object.keys(fields).map(function (fieldKey) {
      return transformFieldToIR(fields[fieldKey]);
    }),
    type: node.type
  };
}

/**
 * Transforms a field (GraphQLInputObjectType or GraphQLScalarType) to a
 * RelayIR ScalarField or LinkedField.
 */
function transformFieldToIR(node) {
  var type = getRawType(node.type);
  if (type instanceof GraphQLInputObjectType) {
    return transformInputObjectToIR(node);
  }

  if (type instanceof GraphQLEnumType || type instanceof GraphQLScalarType) {
    return transformScalarToIR(node.name, type);
  }

  throw new Error('Unhandled node type');
}

/**
 * Transforms a GraphQLScalarType to a RelayIR ScalarField
 */
function transformScalarToIR(name, type) {
  return {
    alias: null,
    args: [],
    directives: [],
    handles: null,
    kind: 'ScalarField',
    metadata: null,
    name: name,
    type: type
  };
}

module.exports = transformInputObjectToIR;
